build_extra/rust/rust.gni

```diff
@@ -47,74 +47,124 @@ rust_test_ldflags = exec_script("get_rust_ldflags.py",
+                           "source_root",
+                           "testonly",
+                           "treat_warnings_as_errors",
+                         ])
+
+  if (!defined(crate_name)) {
+    crate_name = target_name
+  }
+  if (!defined(crate_type)) {
+    crate_type = "rlib"
+  }
+  if (!defined(deps)) {
+    deps = []
+  }
+  if (!defined(is_test)) {
+    is_test = false
+  }
+  if (!defined(libs)) {
+    libs = []
+  }
+  if (!defined(treat_warnings_as_errors)) {
+    # Use global setting if not explicitly specified for this target.
+    treat_warnings_as_errors = rust_treat_warnings_as_errors
+  }
+
+  if (defined(crate_version)) {
+    # In our build setup, all crates are built in the same directory. To avoid
+    # file name conflicts between when multiple versions of the same crate are
+    # built, add a unique suffix to output file names.
+    # Unfortunately the version number as such can't be used directly:
+    # everything after the first dot (.) is thrown away by rust, so in case of
+    # foo-0.2 vs foo-0.3 only the first '0' would be used, and conflicts would
+    # still occur. Therefore we use a hash of the version number instead.
+    crate_suffix = exec_script("//tools/sha256sum.py",
+                               [
+                                 "--input=$crate_version",
+                                 "--format=-%.8s",
+                               ],
+                               "trim string")
+  } else {
+    # Of most crates we use only one version; no need for all this difficulty.
+    crate_suffix = ""
+  }
 
-      # In our build setup, all crates are built in the same directory. To avoid
-      # file name conflicts between when multiple versions of the same crate are
-      # built, add a unique suffix to output file names.
-      #
-      # Unfortunately the version number as such can't be used directly:
-      # everything after the first dot (.) is thrown away by rust, so in case of
-      # foo-0.2 vs foo-0.3 only the first '0' would be used, and conflicts would
-      # still occur. Therefore we use a hash of the version number instead.
-      crate_suffix = exec_script("//tools/sha256sum.py",
-                                 [
-                                   "--input=$crate_version",
-                                   "--format=-%.8s",
-                                 ],
-                                 "trim string")
+  # Derive filenames for 'extern' and 'extern_version' linked rust libraries.
+  extern_rlibs = []
+  if (defined(invoker.extern)) {
+    foreach(label, invoker.extern) {
+      extern_rlibs += [
+        {
+          label = label
+          crate_name = get_label_info(label, "name")
+          rlib = "$out_dir/lib$crate_name.rlib"
+        },
+      ]
+    }
+  }
+  if (defined(invoker.extern_version)) {
+    foreach(info, invoker.extern_version) {
+      extern_rlibs += [
+        {
+          crate_suffix = exec_script("//tools/sha256sum.py",
```

Comment:
Aside: I suspect, without evidence, that this exec_script is slowing incremental builds significantly.

Reply:
Why? It runs in the `gn gen` phase, which takes 2 seconds on my laptop.
Or am I missing something?

Reply:
Also I'd be more than happy to get rid of it if there was a way...